### PR TITLE
Represent holes in Blockly

### PR DESCRIPTION
**Pull request by shmish111**
Deployed to https://david.marlowe.iohkdev.io/

Holes are now represented in Blockly as null (no block)
Notes:
* When going to blockly we lose information about the hole, it just becomes null
* This means that code->blockly->code is not idempotent since the hole names will change

### CODE REVIEW

File: marlowe-playground-client/src/Marlowe/Gen.purs [hunk 1]
```diff
@@ -33,7 +33,13 @@ genRational :: forall m. MonadGen m => MonadRec m => m Rational
...
   n <- genBigInteger
   d <- genBigInteger
-  pure $ Rational n d
+  pure
+    -- this little fix is in Marlowe.Blockly so we need it here to make the tests work
+    
+    $ if d > zero then
+        Rational n d
+      else
+        Rational (-n) (-d)
```
Comment (palas) [inline marlowe-playground-client/src/Marlowe/Gen.purs:42]: Hm, we have an instance of (Eq Rational) that deals with this. Then again, I am guessing some property failed that forced you to add this. It would be best to not have it though, cause users can write any Rational, so there are things we won't be testing.

Does a property fail without this? If so, could you tell me which?

File: marlowe-playground-client/src/Marlowe/Linter.purs [hunk 2]
```diff
@@ -452,7 +452,12 @@ lintValue env t@(Term (Scale h@(Hole _ _ _) c) pos) = do
 
 lintValue env t@(Term (ChoiceValue choiceId a) pos) = do
   modifying _holes (getHoles choiceId)
-  pure (ValueSimp pos false t)
+  sa <- lintValue env a
+  case sa of
+    ConstantSimp _ _ c1 -> pure (ConstantSimp pos true (-c1))
+    _ -> do
+      markSimplification constToVal SimplifiableValue a sa
+      pure (ValueSimp pos false t)
```
Comment (palas) [inline marlowe-playground-client/src/Marlowe/Linter.purs:460]: This is not right, it only works for `NegValue`. But it wasn't right before either, good point. It should always be:
```
sa <- lintValue env a
markSimplification constToVal SimplifiableValue a sa
pure (ValueSimp pos false t)
```
I think. No need to pattern match in `sa`. Maybe I should document this

File: marlowe-playground-client/src/Marlowe/Gen.purs (hunk 1 above)
Comment (shmish111) [reply marlowe-playground-client/src/Marlowe/Gen.purs:42]: This is because of https://github.com/input-output-hk/plutus/blame/master/marlowe-playground-client/src/Marlowe/Blockly.purs#L1414 which makes code->blockly->code non-idempotent, this came up in the tests.

File: marlowe-playground-client/src/Marlowe/Linter.purs (hunk 2 above)
Comment (shmish111) [reply marlowe-playground-client/src/Marlowe/Linter.purs:460]: Ahhh, sorry I misunderstood what was going on here.

**Comment by shmish111** [comment]
> Very nice. I think this is a very convenient functionality, so now you can use both holes and blockly interchangeably whichever is best at each particular point in the editing.
> 
> There is a small glitch with `ChoiceValue`, there is a state where you can have the choice identifier as a hole, and then it will lose whatever you have in the default value when you move to blockly. But you cannot get there by hole expansion. So maybe it would be best to either forbid holes there, or to autocomplete when moving to blockly (so as to not lose information), but it is no big deal, I am happy to leave it as is, maybe we should just get rid of `ChoiceValue` default value...

Following the rest of the implementation, this should not allow holes. This would have come up in tests if they worked for holes which requires having non-unique names. I am going to make that change so the tests should show this up, then I will fix it by not allowing holes.

File: marlowe-playground-client/src/Marlowe/Gen.purs (hunk 1 above)
Comment (palas) [reply marlowe-playground-client/src/Marlowe/Gen.purs:42]: I know, but I mean, I think the tests passed when I wrote this, I fixed it by adding an Eq instance that treats both Rationals the same, so I wonder why it failed now

Comment (palas) [reply marlowe-playground-client/src/Marlowe/Gen.purs:42]: https://github.com/input-output-hk/plutus/blame/3d3b7ac8640204e0919740b5a41b87b86a2b86e3/marlowe-playground-client/src/Marlowe/Semantics.purs#L231

so they should be equal

File: marlowe-playground-client/src/Marlowe/Blockly.purs [hunk 3]
```diff
@@ -962,90 +963,107 @@ caseDefinition g block =
...
+  pure $ Term v { row: 0, column: 0 }
+
+statementToTerm :: forall a. Generator -> Block -> String -> Parser a -> Either String (Term a)
+statementToTerm g block name p = case statementToCode g block name of
+  Either.Left _ -> pure $ Hole name Proxy { row: 0, column: 0 }
+  Either.Right s ->
+    (flip parse) s do
+      v <- p
+      pure $ Term v { row: 0, column: 0 }
+
```
Comment (krisajenkins) [inline marlowe-playground-client/src/Marlowe/Blockly.purs:990]: This is `parse (parseToValue' p) s`, right?

File: marlowe-playground-client/src/Marlowe/Blockly.purs [hunk 4]
```diff
@@ -962,90 +963,107 @@ caseDefinition g block =
...
 
-boundsDefinition :: Generator -> Block -> Either String (Array Bound)
+boundsDefinition :: Generator -> Block -> Either String (Array (Term Bound))
 boundsDefinition g block = traverse (boundDefinition g) (getAllBlocks block)
 
-instance hasBlockDefinitionAction :: HasBlockDefinition ActionType Case where
+parseToValue' :: forall a. Parser a -> Parser (Term a)
+parseToValue' p = do
+  v <- p
+  pure $ Term v { row: 0, column: 0 }
```
Comment (krisajenkins) [inline marlowe-playground-client/src/Marlowe/Blockly.purs:981]: Is this `p >>= mkDefaultTerm`?

File: marlowe-playground-client/src/Marlowe/Holes.purs [hunk 5]
```diff
@@ -205,13 +213,26 @@ constructMarloweType constructorName (MarloweHole { row, column }) m = case Map.
...
 
 mkHole :: forall a. String -> { row :: Pos, column :: Pos } -> Term a
 mkHole name position = Hole name Proxy position
 
+mkDefaultTerm :: forall a. a -> Term a
+mkDefaultTerm a = Term a { row: 0, column: 0 }
+
 data Term a
   = Term a { row :: Pos, column :: Pos }
   | Hole String (Proxy a) { row :: Pos, column :: Pos }
```
Comment (krisajenkins) [inline marlowe-playground-client/src/Marlowe/Holes.purs:238]: This raises a (wholly academic) question, could you give `Term` a `Functor` instance? You'd need a functor instance for `Proxy`, but it's there, so...yes.

Interesting.

File: marlowe-playground-client/src/Marlowe/Gen.purs (hunk 1 above)
Comment (shmish111) [reply marlowe-playground-client/src/Marlowe/Gen.purs:42]: I can't remember why but before this PR it wasn't actually being tested. If you try it out on `master` you will see that `Scale` doesn't work in the playground. Not just this part but the parsing too. This PR makes sure a lot more things are tested and has thrown up quite a few bugs.

File: marlowe-playground-client/src/Marlowe/Blockly.purs (hunk 4 above)
Comment (shmish111) [reply marlowe-playground-client/src/Marlowe/Blockly.purs:986]: no, it's `mkDefaultTerm <$> p`!

File: marlowe-playground-client/src/Marlowe/Holes.purs (hunk 5 above)
Comment (shmish111) [reply marlowe-playground-client/src/Marlowe/Holes.purs:250]: an Applicative might make sense even. But... not right now

File: marlowe-playground-client/src/Halogen/Blockly.purs [hunk 6]
```diff
@@ -92,9 +95,9 @@ handleQuery (Resize next) = do
 handleQuery (SetCode code next) = do
   { blocklyState } <- get
   let
-    contract = case runParser Parser.contractValue code of
+    contract = case Parser.parseContract code of
       Right c -> c
-      Left _ -> Close
+      Left _ -> Hole "root_contract" Proxy { row: 0, column: 0 }
```
[the remark this answers is not in the document]
Comment (shmish111) [reply marlowe-playground-client/src/Halogen/Blockly.purs]: I fixed it as quickly as I could, don't panic

File: marlowe-playground-client/src/Marlowe/Blockly.purs [hunk 7]
```diff
@@ -1264,36 +1289,55 @@ oneCaseToBlockly newBlock workspace (Case (Deposit (AccountId accountNumber acco
   inputToBlockly newBlock workspace block "contract" cont
   pure block
 
-oneCaseToBlockly newBlock workspace (Case (Choice (ChoiceId choiceName choiceOwner) bounds) cont) = do
```
[the remark this answers is not in the document]
Comment (shmish111) [reply marlowe-playground-client/src/Marlowe/Blockly.purs:1267]: In another PR

File: marlowe-playground-client/src/Marlowe/Blockly.purs [hunk 8]
```diff
@@ -1264,36 +1289,55 @@ oneCaseToBlockly newBlock workspace (Case (Deposit (AccountId accountNumber acco
...
+  pure block
+
+oneCaseToBlockly newBlock workspace (Case (Term (Notify observation) _) cont) = do
   block <- newBlock workspace (show NotifyActionType)
   inputToBlockly newBlock workspace block "observation" observation
   inputToBlockly newBlock workspace block "contract" cont
   pure block
 
-nextCase :: forall r. NewBlockFunction r -> STRef r Workspace -> Connection -> Array Case -> ST r Unit
+-- FIXME: This is a temporary hack to get things to compile
```
[the remark this answers is not in the document]
Comment (shmish111) [reply marlowe-playground-client/src/Marlowe/Blockly.purs]: Bigger, non-fixme, comment added

File: marlowe-playground-client/src/Marlowe/Blockly.purs (hunk 3 above)
Comment (shmish111) [reply marlowe-playground-client/src/Marlowe/Blockly.purs]: indeed

File: marlowe-playground-client/src/Marlowe/Gen.purs (hunk 1 above)
Comment (shmish111) [reply marlowe-playground-client/src/Marlowe/Gen.purs:42]: Ah, its not needed now that I got the test to do an extra parse round-trip. Deleting.

Comment (shmish111) [reply marlowe-playground-client/src/Marlowe/Gen.purs:42]: Sigh, after testing a lot a fiddling around I tracked down the issue. We need to do this because in tests where we wrap a `Rational` in a `TermWrapper` when we have two negative values then the column position of the term will different to if we have two positive values, even though the rationals themselves are equal

**Comment by palas** [comment]
🤯, I suspected it had to do with the wrapper, but the column number...
🤦‍♂️
